File: artifacts/api-server/src/integrations/supplier.ts
/**
 * Generic REST supplier adapter.
 * Authenticates via a configurable API-key header.
 * Upserts products from the stock endpoint into the products table.
 */

import type { IntegrationAdapter, SyncResult } from "./types";
import { ZERO_SYNC } from "./types";
import { db, productsTable } from "@workspace/db";
import { logger } from "../lib/logger";
import { withRetry, RateLimiter } from "../lib/rateLimit";

interface SupplierCreds extends Record<string, unknown> {
  baseUrl: string;
  apiKey: string;
  apiKeyHeader?: string;   // default: "X-Api-Key"
  stockPath?: string;      // default: "/inventory"
  listKey?: string;        // JSON key holding the array, empty = bare array
  fieldSku?: string;       // default: "sku"
  fieldStock?: string;     // default: "stock_quantity"
  fieldName?: string;      // default: "name"
  fieldPrice?: string;     // default: "unit_price"
}

function isSupplierCreds(c: Record<string, unknown>): c is SupplierCreds {
  return typeof c.baseUrl === "string" && typeof c.apiKey === "string";
}

const limiter = new RateLimiter(5, 1_000);

async function supplierFetch<T>(creds: SupplierCreds, path: string): Promise<T> {
  await limiter.throttle();
  const url = creds.baseUrl.replace(/\/$/, "") + path;
  const headerName = (creds.apiKeyHeader as string | undefined) ?? "X-Api-Key";
  const res = await fetch(url, {
    headers: { [headerName]: creds.apiKey, "Content-Type": "application/json" },
  });
  if (!res.ok) throw new Error(`Supplier ${res.status}: ${await res.text()}`);
  return (await res.json()) as T;
}

function extractList(data: unknown, listKey: string): unknown[] {
  if (Array.isArray(data)) return data;
  if (data && typeof data === "object" && listKey) {
    const v = (data as Record<string, unknown>)[listKey];
    if (Array.isArray(v)) return v;
  }
  return [];
}

export const supplierAdapter: IntegrationAdapter = {
  platform: "supplier",
  displayName: "Supplier (REST)",
  requiresCredentials: true,

  async validate(credentials) {
    if (!isSupplierCreds(credentials))
      throw new Error("baseUrl and apiKey are required");
    const path = (credentials.stockPath as string | undefined) ?? "/inventory";
    await supplierFetch(credentials, path);
  },

  async sync(userId, credentials) {
    if (!isSupplierCreds(credentials)) throw new Error("Invalid supplier credentials");
    const result: SyncResult = { ...ZERO_SYNC };

    const path = (credentials.stockPath as string | undefined) ?? "/inventory";
    const listKey = (credentials.listKey as string | undefined) ?? "products";
    const fieldSku = (credentials.fieldSku as string | undefined) ?? "sku";
    const fieldStock = (credentials.fieldStock as string | undefined) ?? "stock_quantity";
    const fieldName = (credentials.fieldName as string | undefined) ?? "name";
    const fieldPrice = (credentials.fieldPrice as string | undefined) ?? "unit_price";

    const raw = await withRetry("supplier:stock", () =>
      supplierFetch<unknown>(credentials, path)
    );
    const items = extractList(raw, listKey);

    for (const item of items) {
      const row = item as Record<string, unknown>;
      const sku = String(row[fieldSku] ?? "");
      if (!sku) continue;

      await db
        .insert(productsTable)
        .values({
          userId,
          platform: "supplier",
          externalId: sku,
          name: String(row[fieldName] ?? sku),
          sku,
          category: String(row["category"] ?? "Uncategorized"),
          price: String(Number(row[fieldPrice] ?? 0)),
          cogs: String(Number(row["cost"] ?? row["cogs"] ?? 0)),
          stock: Number(row[fieldStock] ?? 0),
        })
        .onConflictDoUpdate({
          target: [productsTable.userId, productsTable.platform, productsTable.externalId],
          set: {
            stock: Number(row[fieldStock] ?? 0),
            price: String(Number(row[fieldPrice] ?? 0)),
          },
        });
      result.productsAdded += 1;
    }

    logger.info({ userId, products: result.productsAdded }, "Supplier sync ok");
    return result;
  },
};

File: python-api-integration/adapters/supplier.py
"""
adapters/supplier.py
Generic REST supplier adapter (API-key auth).

Your supplier's API must return a JSON list or a JSON object with a list
under a known key. Set SUPPLIER_LIST_KEY (default "products") if needed.
"""

from __future__ import annotations

import logging
import os
from typing import Any

from .base import BaseAdapter, _retryable
from credentials import SupplierCreds

logger = logging.getLogger(__name__)

# The JSON key that contains the product list in the supplier response.
# If the response IS the list (a bare array), set this to an empty string.
_LIST_KEY = os.getenv("SUPPLIER_LIST_KEY", "products")

# Endpoint path for stock (e.g. "/inventory" or "/stock")
_STOCK_PATH = os.getenv("SUPPLIER_STOCK_PATH", "/inventory")

# JSON field mappings — override via env vars to match your supplier schema
_FIELD_SKU = os.getenv("SUPPLIER_FIELD_SKU", "sku")
_FIELD_STOCK = os.getenv("SUPPLIER_FIELD_STOCK", "stock_quantity")
_FIELD_NAME = os.getenv("SUPPLIER_FIELD_NAME", "name")
_FIELD_PRICE = os.getenv("SUPPLIER_FIELD_PRICE", "unit_price")


class SupplierAdapter(BaseAdapter):
    """Fetch stock levels from a generic REST supplier API."""

    _rate_limit_calls = 5
    _rate_limit_period = 1.0

    def __init__(self, creds: SupplierCreds | None = None) -> None:
        super().__init__()
        self._creds = creds or SupplierCreds()

    def _auth_headers(self) -> dict[str, str]:
        return {self._creds.api_key_header: self._creds.api_key}

    # ---------------------------------------------------------------------- #
    # Public interface                                                          #
    # ---------------------------------------------------------------------- #
    @_retryable
    def get_stock(
        self,
        path: str = _STOCK_PATH,
        params: dict | None = None,
    ) -> list[dict[str, Any]]:
        """
        Fetch raw stock data from the supplier API.

        Args:
            path:   API path to call (appended to SUPPLIER_BASE_URL).
            params: Optional query parameters.

        Returns:
            List of raw product dicts.
        """
        url = self._creds.base_url.rstrip("/") + path
        data = self._get(url, params=params or {})

        # Handle both bare-list and wrapped-object responses
        if isinstance(data, list):
            items = data
        elif isinstance(data, dict):
            items = data.get(_LIST_KEY, []) if _LIST_KEY else []
            if not items and not _LIST_KEY:
                # Treat entire dict as single product
                items = [data]
        else:
            items = []

        logger.info("Supplier: fetched %d stock lines", len(items))
        return items

    def sync_supplier_stock(
        self,
        path: str = _STOCK_PATH,
        params: dict | None = None,
    ) -> list[dict[str, Any]]:
        """
        Normalised output for the orchestrator.

        Returns:
            [
                {
                    "sku": str,
                    "name": str,
                    "stock": int,
                    "price": float,
                    "source": "supplier",
                },
                ...
            ]
        """
        raw = self.get_stock(path=path, params=params)
        return [
            {
                "sku": str(item.get(_FIELD_SKU, "")),
                "name": item.get(_FIELD_NAME, ""),
                "stock": int(item.get(_FIELD_STOCK, 0)),
                "price": float(item.get(_FIELD_PRICE, 0)),
                "source": "supplier",
            }
            for item in raw
        ]
